### ISSUE TITLE
OpenKeyChain fails as encryption provider for K9-Mail with a `NullPointerException` in `selectUserIdStatusByEmailLike`

### ISSUE THREAD

**Issue by pokulo** [open-keychain/open-keychain#2881]
OpenKeyChain fails as encryption provider for K9-Mail with a `NullPointerException` in `selectUserIdStatusByEmailLike`.

## Expected Behavior
If the entered contact name or mail address (or text) does not match any pgp key owner, then simply hide the autocomplete
pop-up. To avoid the crash, OpenKeyChain should not raise a `NullPointerException`, but return `null`.

## Current Behavior
k9-mail crashes completely upon entering contact name to new email editor. `adb logcat` reveales the following traceback:

```java
02-12 10:45:19.578 13611 13611 D TokenAutoComplete: Current completion text: XXXXX,
02-12 10:45:19.598  8074  8103 E DatabaseUtils: Writing exception to parcel
02-12 10:45:19.598  8074  8103 E DatabaseUtils: java.lang.NullPointerException: ResultSet returned null for UserPackets.sq:selectUserIdStatusByEmailLike
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at com.squareup.sqldelight.Query.executeAsOne(SourceFile:16)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at org.sufficientlysecure.keychain.daos.UserIdDao.getUidStatusByEmailLike(SourceFile:13)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at org.sufficientlysecure.keychain.remote.KeychainExternalProvider.loadUidStatusMap(SourceFile:14)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at org.sufficientlysecure.keychain.remote.KeychainExternalProvider.query(SourceFile:246)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at android.content.ContentProvider.query(ContentProvider.java:1455)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at android.content.ContentProvider.query(ContentProvider.java:1551)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at android.content.ContentProvider$Transport.query(ContentProvider.java:285)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:107)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at android.os.Binder.execTransactInternal(Binder.java:1280)
02-12 10:45:19.598  8074  8103 E DatabaseUtils: 	at android.os.Binder.execTransact(Binder.java:1244)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: FATAL EXCEPTION: ModernAsyncTask #1
02-12 10:45:19.600 13611 15074 E AndroidRuntime: Process: com.fsck.k9, PID: 13611
02-12 10:45:19.600 13611 15074 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at androidx.loader.content.ModernAsyncTask$3.done(ModernAsyncTask.java:164)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:381)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:250)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:269)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:1012)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: Caused by: java.lang.NullPointerException: ResultSet returned null for UserPackets.sq:selectUserIdStatusByEmailLike
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:3017)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:2995)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2978)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:190)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:142)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.content.ContentProviderProxy.query(ContentProviderNative.java:481)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.content.ContentResolver.query(ContentResolver.java:1219)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.content.ContentResolver.query(ContentResolver.java:1151)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at android.content.ContentResolver.query(ContentResolver.java:1107)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at com.fsck.k9.activity.compose.RecipientLoader.fillContactDataFromCryptoProvider(RecipientLoader.java:232)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at com.fsck.k9.activity.compose.RecipientLoader.loadInBackground(RecipientLoader.java:202)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at com.fsck.k9.activity.compose.RecipientLoader.loadInBackground(RecipientLoader.java:37)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at androidx.loader.content.AsyncTaskLoader.onLoadInBackground(AsyncTaskLoader.java:307)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at androidx.loader.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:60)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at androidx.loader.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:48)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at androidx.loader.content.ModernAsyncTask$2.call(ModernAsyncTask.java:141)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
02-12 10:45:19.600 13611 15074 E AndroidRuntime: 	... 3 more
02-12 10:45:19.607  1776 15107 I DropBoxManagerService: add tag=data_app_crash isTagEnabled=true flags=0x2
02-12 10:45:19.608  1776  2987 W ActivityTaskManager:   Force finishing activity com.fsck.k9/.activity.MessageCompose
02-12 10:45:19.610  1776  2987 W ActivityTaskManager:   Force finishing activity com.fsck.k9/.activity.MessageList
```

## Possible Solution
Since the dependency upgrade in commit: 5d84bd838744f63d46d5a08110bad13b5192ae6a the implementation of the method `getUidStatusByEmailLike` was changed like this:
```diff
    public UidStatus getUidStatusByEmailLike(String emailLike) {
-        SqlDelightQuery query = UserPacket.FACTORY.selectUserIdStatusByEmailLike(emailLike);
-        return mapSingleRow(query, UserPacket.UID_STATUS_MAPPER);
+        return getDatabase().getUserPacketsQueries().selectUserIdStatusByEmailLike(emailLike)
+                .executeAsOne();
    }
```

The custom method [`mapSingleRow`](https://github.com/open-keychain/open-keychain/commit/5d84bd838744f63d46d5a08110bad13b5192ae6a#diff-4e64a2dc3818bf32742bd9a297a1712100ab165ea16f86fa98ef5f1b13a270c0L56) did simply return `null`, if no row matched the `emailLike`. But the now used method  [`sqldelight.Query.executeAsOne`](https://github.com/cashapp/sqldelight/blob/99eec4d8552d3729a7f749db529ddc371c488090/runtime/src/commonMain/kotlin/app/cash/sqldelight/Query.kt#L187) simply raises fatal `NullPointerException` as documented. I suggest using the method [`executeAsOneOrNull`](https://github.com/cashapp/sqldelight/blob/99eec4d8552d3729a7f749db529ddc371c488090/runtime/src/commonMain/kotlin/app/cash/sqldelight/Query.kt#L198C7-L198C25) instead to restore the old behavior:

```diff
    public UidStatus getUidStatusByEmailLike(String emailLike) {
        return getDatabase().getUserPacketsQueries().selectUserIdStatusByEmailLike(emailLike)
+                .executeAsOne();
+                .executeAsOneOrNull();
    }
```

## Steps to Reproduce
1. Configure OpenKeyChain as crypto-provider to k9-mail
2. Open "Compose Mail" to draft a new mail
3. Enter some initial text (like "XXXXX" in the traceback) to the mail-address field (To:) to start email-autocompletion

## Context
I want to write (encrypted) emails using k9-mail+OKC.

## Your Environment
* Android Version: Android 13 (LineageOS 20-20240114-microG-sagit)
* Device Model: Xiaomi Mi6 (sagit)
* OpenKeychain Version: 6.0.0
* From F-Droid
* K9-Mail version: 6.603 (from F-Droid)

**Comment by q-pa** [comment]
Can confirm crash with k9mail when filling any address field in email.

**Comment by basilgello** [comment]
I built OKC with your patch and I can confirm no crash in k9mail now! Thanks!

**Comment by Valodim** [comment]
fixed in #2882, will be in the next version

**Comment by kolAflash** [comment]
For reference: https://gitlab.com/fdroid/fdroiddata/-/issues/3197

**Comment by pokulo** [comment]
My bugfix in PR #2882 apparently fell short in fixing the modified method completely.  There is no `NullPointerException: ResultSet returned null for UserPackets.sq:selectUserIdStatusByEmailLike`, since empty results are allowed now. I did not consider the other case: The same query `selectUserIdStatusByEmailLike` can also return multiple results, which is disallowed in a similar manner by both `executeAsOneOrNull` and `executeAsOne` and results in the following error and the same crash of K9-Mail:

```
02-25 20:43:50.630 23142 23157 E DatabaseUtils: java.lang.IllegalStateException: ResultSet returned more than 1 row for UserPackets.sq:selectUserIdStatusByEmailLike
02-25 20:43:50.630 23142 23157 E DatabaseUtils: 	at com.squareup.sqldelight.Query.executeAsOneOrNull(SourceFile:48)
02-25 20:43:50.630 23142 23157 E DatabaseUtils: 	at org.sufficientlysecure.keychain.daos.UserIdDao.getUidStatusByEmailLike(SourceFile:13)
```

I did not dig deeper for a solution. But I propose to reopen this Issue, because it is almost the same error. I can also open another fresh issue, if that's better.
